scripts/gis-material-update/process/modules/hsl.py

```diff
@@ -1,14 +1,381 @@
+        ]
+        stop_times_trip_uniq = stop_times_trip.drop_duplicates("trip_id")
+        stops_trips = stop_times_trip_uniq.merge(trips_day, on="trip_id", how="left")
+
+        # rush hour time columns added
+        stops_trips = self._compute_rush_hours(stops_trips)
+
+        shape_trips_max = (
+            stops_trips.loc[
+                :, ["shape_id"] + [x for x in stops_trips.columns if x.startswith("n_")]
+            ]
+            .groupby("shape_id")
+            .sum()
+            .max(axis=1)
+        )
+        stop_times_day = shape_trips_max.to_frame("rush_hour").reset_index()
+
+        # pick weeks worth of service_ids
+        w_start = self.transit_week()[0]
+        w_end = self.transit_week()[-1]
+
+        ind_service_ids_week = self._pick_service_ids(w_start, w_end)
+        service_ids_week = (
+            self.feed().calendar[ind_service_ids_week]["service_id"].values.tolist()
+        )
+
+        ind_trips_week = self.feed().trips["service_id"].isin(service_ids_week)
+        trips_week = self.feed().trips[ind_trips_week]
+
+        shps = self.feed().shapes.copy()
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=ShapelyDeprecationWarning)
+            # form point objects from coordinate values
+            shps["geometry"] = shps.apply(
+                lambda r: Point(r.shape_pt_lon, r.shape_pt_lat), axis=1
+            )
+
+        # pick all line geometries (shapes) that are operated within a given week
+        # sort by shape_id and shape point sequence, to ensure correct linestring formation
+        shapes_week_ind = shps["shape_id"].isin(trips_week["shape_id"].tolist())
+        shapes_week = shps[shapes_week_ind].sort_values(
+            ["shape_id", "shape_pt_sequence"]
+        )
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=ShapelyDeprecationWarning)
+            # Create Linestring from points
+            shapes_week_lines = gpd.GeoDataFrame(
+                shapes_week.groupby(["shape_id"])["geometry"].apply(
+                    lambda x: LineString(x.tolist())
+                ),
+                geometry="geometry",
+            )
+
+        shapes_week_lines = shapes_week_lines.merge(
+            stop_times_day, on="shape_id", how="left"
+        ).fillna(0)
+
+        # merge shapes and routes
+        trips_week = trips_week.drop_duplicates("shape_id")
+        shapes_with_trips = shapes_week_lines.merge(
+            trips_week, on="shape_id", how="left"
+        )
+        shapes_trips_routes = shapes_with_trips.merge(
+            self.feed().routes, on="route_id", how="left"
+        )
+        shapes_trips_routes = shapes_trips_routes[~shapes_trips_routes.route_id.isna()]
+
+        # Exclude route types
+        # 0 - trams
+        # 1 - subway
+        # 4 - Suomenlinna ferries
+        # 109 - trains
```

Comment:
Would be possible to whitelist the route types for buses, i.e. the types we are interested in? In general, whitelisting is more robust than blacklisting. Like, we don't want to include the Raide-Jokeri here if they decide to use route type like "903 Regional Tram Service".

Reply:
To be done in https://helsinkisolutionoffice.atlassian.net/browse/HAI-1327